### PR TITLE
Implementing new UX for invoking rich text Link UI 

### LINKED ISSUE #57821: Improving the UX around invoking the LinkControl popover
I'm exploring ideas for improving how LinkControl functions within RichText, particularly around invoking the LinkControl popover. I wanted to bring up this approach that Slack uses for links, that I think is interesting. 

### Current

The current implementation in the editor: 
- Uses the block toolbar button to create a link, if a link is not selected. 
- Uses the block toolbar button to "unlink" a link, if a link is selected. 
- Invokes the LinkControl popover, if you select/click a link. 
- Invokes the LinkControl popover, if you use the keyboard to move your cursor onto the link. 

Here's a video observing those behaviors:

https://github.com/WordPress/gutenberg/assets/1813435/dbd19f74-ddc0-44dc-9f71-160eec52e122

### Proposed

The proposed implementation: 
- Uses the block toolbar button to create a link, if a link is not selected (_no change_).
- Uses the block toolbar button to invoke the LinkControl popover, if a link is selected (_change_).
- Invokes the LinkControl popover, if you select/click a link (_no change_). 
- Does not invoke the LinkControl popover, if you use the keyboard to move your cursor onto the link (_change_). 

Here's a video observing those behaviors, as seen in the Slack editor: 

https://github.com/WordPress/gutenberg/assets/1813435/636eae2b-7aaa-43c9-9db7-8775177c0b35

### What I like about this approach:

- The link/unlink control in the toolbar is never destructive; it always invokes the LinkControl popover (either to create, or edit).
- In the current implementation, we have two "unlink" controls always available at the same time. One in the toolbar, when you select a link. The other in the LinkControl popover, which is always available when you select the link anyhow. Related: https://github.com/WordPress/gutenberg/pull/57775.
- It's generally more predictable. 

### Why

This is based on accessibility feedback that link editing is not easily discoverable, and UX feedback that the popover can be distracting opening whenever the cursor enters the link:

> For a screen reader user, IMO there isn't a discoverable way to edit a link. The toolbar includes an Unlink button, but no Edit Link button. To edit a link they'd need to know the Command + K convention (a common pattern, which is great) or know to press Tab when the carat is on the linked text.

Related comment by @jeryj on https://github.com/WordPress/gutenberg/pull/57726#issuecomment-1889652287

> One more note and it has been discussed in previous issues but never fixed. When editing a link, you must use the shortcut. From what I can tell, there is only an unlink button in the formatting toolbar and there should also likely be an edit option there for best discovery. It makes little sense that by highlighting text, I would then tab to find a random edit button for my created link when tabbing under any other condition would land me in the sidebar.

Related comment by @alexstine on https://github.com/WordPress/gutenberg/pull/54063#issuecomment-1708858002:

### PR DESCRIPTION
**Pull request by getdave**
## What?

Changes UX for links within rich text to require explicit activation before displaying the Link UI interface.

Closes https://github.com/WordPress/gutenberg/issues/57821

## Why?

Currently on `trunk` moving the cursor within an existing link (format) will cause the Link UI to appear. This is providing a sub par a11y experience because:

- there is nothing to alert users of assistive tech that the Link UI has appeared
- it disrupts the standard tab order of the editor because pressing tab when a link is active will cause focus to shift to the link popover

## How?

- disables the behaviour to immediately activate the Link UI is the link format is active
- adds a click handler to the contenteditable element which will trigger the Link UI when it is clicked.
- changes the block toolbar button to be "Edit link" if the link format is active. Clicking this will cause the Link UI to appear.

- New Post
- Create a paragraph block
- Add some text
- Create 3 or 4 links
- Click on each link and see you can activate and edit the link using the Link UI.
- Using keyboard attempt the same thing but notice that you must either click the toolbar button or press `CMD + K` to edit the link

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/444434/02ca32c4-fbba-461c-a23f-039d6b73a8b7

### CODE REVIEW

File: packages/format-library/src/link/index.js
Comment (jeryj) [inline packages/format-library/src/link/index.js]: Instead of managing a `clickedLink` state, can we set the state of the thing we want to do? I.e. - display the link preview.

**Review by jeryj — COMMENTED** [review]
I have some of the other work for this issue I can push up to this branch.

File: packages/format-library/src/link/index.js
Comment (ellatrix) [inline packages/format-library/src/link/index.js]: Can't you just return early?

Comment (ellatrix) [reply packages/format-library/src/link/index.js]: Also if anchorElement is defined and isActive is true, it's guaranteed to be a link no?

**Review by ellatrix — APPROVED** [review]
Sure, let's try.

**Comment by getdave** [comment]
> Sure, let's try.

Thank you Ella. 

@jeryj Let's refine the PR UX / UI before this gets merged

File: packages/format-library/src/link/index.js
Comment (getdave) [reply packages/format-library/src/link/index.js]: No because it can be a VirtualAnchorElement returned from `useAnchor`.

I've ended up avoiding the `useAnchor` entirely now by attaching listeners to the contenteditable.

File: packages/format-library/src/link/index.js
Comment (getdave) [reply packages/format-library/src/link/index.js]: I don't see how we could do that. The `Popover` is within the React tree. Therefore we must setState to trigger a re-render of the component with the new state.

**Comment by getdave** [comment]
✅ Update: I implemented this in https://github.com/WordPress/gutenberg/pull/57986/commits/a8a418a183071ff45e7d95b7fb43039c2df00824

------

I think the next step should be that if you are within a link and...

- press Cmd + K
- press the 'Edit link` button in the block toolbar

...then 

- the Link UI appears
- the Link UI is given focus
- the Link UI is in "preview" mode (as opposed to the "edit" mode form as it is currently)

I think we can safely trap focus and constrain tabbing.

We should then consider incorporating the a11y tweaks to labelling the dialog that I started in https://github.com/WordPress/gutenberg/pull/54063.

File: packages/format-library/src/link/index.js
Comment (jeryj) [reply packages/format-library/src/link/index.js]: Can we use `setAddingLink( true )` instead of adding `clickedLink` state?

File: packages/format-library/src/link/index.js
Comment (jeryj) [reply packages/format-library/src/link/index.js]: > No because it can be a VirtualAnchorElement returned from useAnchor.

How can I test VirtualAnchorElements? What blocks use them?

Comment (getdave) [reply packages/format-library/src/link/index.js]: You can't 😄 Take a look at `useAnchor` and you'll see it returns a "virtual" anchor representing a selection within RichText. So for example, when you highlight some text and you want to make a link, the Link UI needs to have an anchor point to "attach" the Popover to, but the link (`<a>` TAG) doesn't yet exist in the DOM. So `useAnchor` will return a coordinate which approximates the selection in the richtext.

File: packages/format-library/src/link/index.js
Comment (getdave) [reply packages/format-library/src/link/index.js]: Yeh try that. Might not need two states.

Comment (jeryj) [reply packages/format-library/src/link/index.js]: Done :) https://github.com/WordPress/gutenberg/pull/57986/commits/b322bbd5f442672969f7b988ffa32966d0393845

File: packages/format-library/src/link/index.js
Comment (getdave) [reply packages/format-library/src/link/index.js]: Nice work!

We'll need some new e2e test coverage for these changes. Specifically it would be super easy for regressions to be introduced around this new behaviour so we need to validate the new expectations.

**Comment by richtabor** [comment]
I suppose the link control in the toolbar should open/close the LinkControl, just like when the other formats are applied (and not subsequently removed upon a second click): 

https://github.com/WordPress/gutenberg/assets/1813435/666eafb5-37b8-49bb-baae-8cc587690859

**Comment by jeryj** [comment]
> I suppose the link control in the toolbar should open/close the LinkControl, just like when the other formats are applied (and not subsequently removed upon a second click):

I've been looking at this, and it is unfortunately quite tricky due to how/when states are set. Are you OK with that behavior in a follow-up?

**Comment by richtabor** [comment]
> Are you OK with that behavior in a follow-up?

Sure.

File: packages/format-library/src/link/index.js
Comment (jeryj) [reply packages/format-library/src/link/index.js]: I think fixing all the broken e2e tests will validate it too.

**Comment by getdave** [comment]
Update: I can't replicate this once I rebased and then reverted https://github.com/WordPress/gutenberg/pull/57986/commits/74a3e9ef2c06801608ac54627bd96e5dcd52f61a

-------

@jeryj I found a bug where we show the wrong UI:

- Add text
- Create link and submit
- Immediately click link again
- See the "creating link" state rather than the preview of the existing link.


https://github.com/WordPress/gutenberg/assets/444434/08982624-e877-4b2e-8b38-5562083f30c9


It only seems to happen if you click exactly on the point where the cursor is placed.

**Comment by getdave** [comment]
Rebased to bring in changes made in https://github.com/WordPress/gutenberg/pull/57726

**Comment by richtabor** [comment]
I noticed an issue when clicking on the link control in the toolbar, then interacting elsewhere, if the preview is active: 


https://github.com/WordPress/gutenberg/assets/1813435/be1b62fc-da9e-4b00-accf-51ff786de3f8

**Comment by jeryj** [comment]
> See the "creating link" state rather than the preview of the existing link.

Showing the create link vs preview link seems to be a bug with _state_ of the popover and not the state of if the popover should be open or not. It seems like only using one `addingLink` state is able to handle the "Is the popover opened or closed" correctly. I debugged this more, and it only happens if you flip the state of "Open in a new tab," so I don't think this bug is related to using `addingLink` to handle the popover, but a deeper issue. 

I tested this same flow on `trunk`, and if you change the state of "Open in a new tab," and click, it isn't clickable for the first few times as some kind of state gets updated before it realizes it's a link again. 

> I noticed an issue when clicking on the link control in the toolbar, then interacting elsewhere, if the preview is active:

I believe that issue is fixed if we don't use the setClickState. Or, rather, we need to check if the element is a link or not.

**Comment by getdave** [comment]
@jeryj Apologies. I seem to have destroyed some of your hard work. I genuinely thought I'd reverted to fix stuff but I realised your commits had some additional changes I neglected to restore 🙏

**Comment by getdave** [comment]
https://github.com/WordPress/gutenberg/assets/444434/f47fd7d2-4523-4601-bded-e953a90ecf0f

Going back to @jeryj's code seems to have fixed Rich's issue. But there is a new Issue whereby the `addingLink` state is set for a click on an existing link's `<a>` tag but when that link format is not _active_. As result the `activeAttributes` are not populate and thus `LinkControl` shows the "creating" mode.

Fixing that now.

**Comment by jeryj** [comment]
This is ready for functional review, but we do need more tests to solidify the flows here. The big improvement now is that pressing ESCAPE will return you to whatever opened the popover. It's pretty smooth on the keyboard.

I have not figured out how to have the link or toolbar edit link button toggle the state of the popover, as `onFocusOutside` fires before the click event:
- Link control is open
- Click on Toolbar Edit link button to close it
- `onFocusOutside` fires, closing the popover
- Click event on the Toolbar Edit link button fires, re-opening the link control, when we were intending to close it.

File: packages/format-library/src/link/index.js
Comment (jeryj) [inline packages/format-library/src/link/index.js:43]: I didn't want to add this `openedBy` state, as the popover has its own system for returning focus. For some reason, it's not working correctly with the Rich text element and split between iframe editor and toolbar in the header.

**Comment by getdave** [comment]
> This is ready for functional review, but we do need more tests to solidify the flows here. The big improvement now is that pressing ESCAPE will return you to whatever opened the popover. It's pretty smooth on the keyboard.

Great work @jeryj 👏 

> I have not figured out how to have the link or toolbar edit link button toggle the state of the popover, as `onFocusOutside` fires before the click event:


@richtabor [already confirmed he's happy with this in a followup](https://github.com/WordPress/gutenberg/pull/57986/#issuecomment-1906714093). I don't think we should block this PR further on this unless there's a very important reason why it has to be in the initial release.

**Review by scruffian — APPROVED** [review]
This is testing well for me.

**Comment by jeryj** [comment]
Tests are hopefully fixed to account for the new states of the Link Control being visible/hidden. 🤞🏻 Also added tests to cover [focus outside click sending focus to the right area](https://github.com/WordPress/gutenberg/pull/57986/commits/783a37823ade226f18f00dbc42cc51f980c35fbf) and returning focus from the keyboard to the button that opened it.

**Review by richtabor — APPROVED** [review]
LGTM ✨

**Pull request merged**
